executor/show.go

```diff
@@ -1858,8 +1859,55 @@ func (e *ShowExec) fetchShowTableRegions() error {
 		regions, err = getTableRegions(tb, physicalIDs, tikvStore, splitStore)
 	}
 
-	if err != nil {
-		return err
+	// not for table index
+	if len(e.IndexName.L) == 0 {
+		var scheduleState map[int64]infosync.PlacementScheduleState
```

Comment:
The problem is that placement rules has been released. Our model may have been persisted. Thus if we want to add one, it may not be zero.

I think we could try to initialize every table/partition. Or ensure by an extra `map[xxx]bool`?

Reply:
> The problem is that placement rules has been released. Our model may have been persisted. Thus if we want to add one, it may not be zero.
> 
> I think we could try to initialize every table/partition. Or ensure by an extra `map[xxx]bool`?

So some where stored the `PlacementScheduleState` as a persisted value?

Reply:
After some digging, I think no struct are referfing `PlacementScheduleState`. Thus we could add a new enumeration.

Reply:
Yep it looks suspicious indeed. 
Actually pending state is from line 60, because etcdCli is nil. Otherwise it'll fetch from PD. https://github.com/pingcap/tidb/blob/06737ecd29b0ae633d9ac89b0660e7a43939c88d/domain/infosync/region.go#L53-L69
Currently it reads from PD, @xhebox could you please explain a little bit more about "add a new enumeration"?

Reply:
OK. The thing is that, `PlacementScheduleStatePending == 0`. So \ `var scheduleState map[int64]infosync.PlacementScheduleState` will get `scheduleState[xx] == PlacementScheduleStatePending` by default for any `xx` that is not fetched from PD.

> Because PlacementScheduleState.Pending is 0, so get value from an nil scheduleState will always return PlacementScheduleState.Pending , the test cases still passed even though the code is not right.

Reply:
Hi @xhebox, thanks for the reply! I think what you mentioned is's a pervasive issue that before this change. A typical usage of `fetchScheduleState` and `scheduleState` is:
https://github.com/pingcap/tidb/blob/2e1cff73f3bf7ee18d498a04138d39d70bd85a6d/executor/show_placement.go#L404-L415
In `fetchAllTablePlacements` function, `fetchScheduleState ` is called several times to fetch state from PD. `if block` at line 405 servers as a cache so that we don't need to ping PD if there's a match. But for key not found in `scheduleState`, `ok` will be false then it will go to line 410, call `GetReplicationState` to fetch from PD. I verified that in unit test, the pending state is returned from line 60 in `GetReplicationState ` function, not from the map. Please take a second look.

One more thing is we return pending state as a default state, even when there's something wrong with the system, which makes it error prone. But on a second thought, there's a `in progress` state, so call it pending still make sense. Or we can add another state to indicate if there's any system error cause the scheduling issue. But again, this is how states works before this change, if we think it's better to add another state, I think maybe I should start another PR, since there's many other places needs to update.

Reply:
> But again, this is how states works before this change, if we think it's better to add another state, I think maybe I should start another PR, since there's many other places needs to update.

That makes sense. I agreed to leave it to the next PR, @lcwangchao , PTAL

Reply:
Thanks! Created a new issue here: https://github.com/pingcap/tidb/issues/35314, please feel free to let me know if we want to update the state